### PR TITLE
Build script changes for SxS willow instances

### PR DESCRIPTION
**Pull request by alpaix**
Sharing build script changes to support development in SxS willow environment.

`configure.ps1` script is now able to identify the latest willow instance and deduce VS install dir and msbuild path. Configuration parameters are store in `artifacts\configure.json` file so that build or test script don't need to run in the same PS session.

Fixed VC++ project system interop assembly dependency.

Optional. Set VSIX version in manifest to 3.6.99.99 to be able to publish locally built extension to experimental instance. Otherwise this fails due to version conflict. This change is safe as CI overwrites the version string in the manifest on S&P build step.

//cc @rohit21agrawal @emgarten @joelverhagen @drewgil @rrelyea @jainaashish

### CODE REVIEW

File: build/common.ps1 [hunk 1]
```diff
@@ -311,7 +322,7 @@ function Enable-DelaySigningForDotNet {
 
 Function Save-ProjectFile ($xproject, $fileName) {
     Trace-Log "Saving project to '$fileName'"
-    $xproject | ConvertTo-Json -Depth 100 | Out-File $fileName
+    $xproject | ConvertTo-Json -Depth 99 | Out-File $fileName
```
Comment (joelverhagen) [inline build/common.ps1]: 100 is an acceptable number (the max), why the change?

File: build/common.ps1 [hunk 2]
```diff
@@ -257,28 +253,43 @@ Function Test-MSBuildVersionPresent {
...
+         %{ Set-Item -Path "env:$($_.Name)" -Value $_.Value }
+
     if ($CI) {
         # Explicitly add cli to environment PATH
         # because dotnet-install script runs in configure.ps1 in previous build step
         $env:path = "$CLIRoot;${env:path}"
     }
 }
 
+Function Use-BuildToolset {
```
Comment (joelverhagen) [inline build/common.ps1]: What's this empty function for?

File: configure.ps1 [hunk 3]
```diff
@@ -34,50 +31,139 @@ Param (
...
+} -ev +BuildErrors
 
-Install-DotnetCLI -Force:$Force
+Invoke-BuildStep 'Installing .NET CLI' {
+    Install-DotnetCLI -Force:$Force
+} -ev +BuildErrors
 
-if ($CleanCache) {
+# Restoring tools required for build
+Invoke-BuildStep 'Restoring solution packages' {
```
Comment (joelverhagen) [inline configure.ps1:51]: This should be done in `build.ps1` still. There are packages used by the build, i.e. `Microsoft.VisualStudio.ProjectSystem`. Additionally, all the other restores are happening there 😄

File: configure.ps1 [hunk 4]
```diff
@@ -34,50 +31,139 @@ Param (
...
+    BuildTools = @{}
+    Toolsets = @{}
 }
 
-Trace-Log "Validating VS${ToolsetVersion} toolset installation"
+Function New-BuildToolset {
+    param(
+        [int]$ToolsetVersion
+    )
+    $CommonToolsVar = "Env:VS${ToolsetVersion}0COMNTOOLS"
```
Comment (joelverhagen) [inline configure.ps1:68]: This will be set in the Developer Command Prompt right? Is that a requirement? Or will this stuff also work from a vanilla PowerShell window?

**Review by joelverhagen — COMMENTED** [review]
Can you further explain the VSIX version change?

File: build/common.ps1 (hunk 2 above)
Comment (alpaix) [reply build/common.ps1]: A placeholder for a method to switch toolsets between build steps, e.g for setting environment variables. Will remove for now.

File: build/common.ps1 (hunk 1 above)
Comment (alpaix) [reply build/common.ps1]: rebase issue. Will fix.

File: configure.ps1 (hunk 3 above)
Comment (alpaix) [reply configure.ps1:51]: There are more build tools I'd wanted to validate and configure here. Project system packages should leave the `packages.config` eventually.

File: configure.ps1 (hunk 4 above)
Comment (alpaix) [reply configure.ps1:68]: Classic VS install sets this variable for current user or machine. Given that for back-compat I check it first. If it's not set I check the registry keys, then search for willow instances. If none of the above succeeds - assume the default location.

No need to pre-set it. This works for me even in a non-dev command prompt.

**Comment by alpaix** [comment]
@joelverhagen VS "15" comes with pre-installed Package Manager v3.6.0. Hence deploying locally built VSIX to an experimental instance of the same version will fail due to version conflict error. VS requires the version to be higher than the pre-installed one. I made the manifest version to be sufficiently high to override "release" version and to be easily distinguished from "official" build. Does this make sense to you?

**Comment by joelverhagen** [comment]
Yup. Makes sense. Can we make this 99.99.99 so we never have to change it again?

**Comment by alpaix** [comment]
Initially I thought to preserve major and minor as a hint. 
Well, after thinking this through again, it sound like a good idea to set it once and forever.

File: configure.ps1 (hunk 3 above)
Comment (joelverhagen) [reply configure.ps1:51]: Ah makes sense. Perhaps we could restore in both places until we have finished removing product dependencies from this `packages.config`?

File: build.ps1 [hunk 5]
```diff
@@ -123,54 +113,39 @@ Invoke-BuildStep 'Cleaning artifacts' {
...
     -skip:$SkipVS15 `
     -ev +BuildErrors
 
 ## Building the VS14 Tooling solution
 Invoke-BuildStep 'Building NuGet.Clients projects - VS14 Toolset' {
-        param($Configuration, $ReleaseLabel, $BuildNumber, $SkipRestore)
-        Build-ClientsProjects $Configuration $ReleaseLabel $BuildNumber -ToolsetVersion 14 -SkipRestore:$SkipRestore
+        Build-ClientsProjects $Configuration $ReleaseLabel $BuildNumber -ToolsetVersion 14
     } `
-    -args $Configuration, $ReleaseLabel, $BuildNumber, $SkipRestore `
```
Comment (joelverhagen) [inline build.ps1:161]: Nice cleanup 👍

File: src/NuGet.Clients/PackageManagement.VisualStudio/NuGet.PackageManagement.VisualStudio.csproj [hunk 6]
```diff
@@ -55,15 +43,19 @@
     </When>
   </Choose>
   <ItemGroup>
+    <Reference Include="envdte, Version=8.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a">
+      <EmbedInteropTypes>False</EmbedInteropTypes>
```
Comment (joelverhagen) [inline src/NuGet.Clients/PackageManagement.VisualStudio/NuGet.PackageManagement.VisualStudio.csproj]: `<EmbedInteropTypes>` is false by default, right?

Comment (alpaix) [reply src/NuGet.Clients/PackageManagement.VisualStudio/NuGet.PackageManagement.VisualStudio.csproj]: It used to be that way before my change. And I don't want to touch this.

File: configure.ps1 (hunk 3 above)
Comment (alpaix) [reply configure.ps1:51]: Actually this is about to happen very soon when we start consuming new CPS API. Also I'd wanted to clean up the build script from all one-time setup steps now.

**Review by joelverhagen — APPROVED** [review]
